netket/utils/history.py

```diff
@@ -12,205 +12,195 @@
 
-        elif hasattr(values, "to_compound"):
-            _value_name, value_dict = values.to_compound()
-            _value_dict = {}
-            _len = 1
-            for (key, val) in value_dict.items():
-                _value_dict[key] = np.array([val], dtype=dtype)
-                _keys.append(key)
-
-        if iters is None:
-            if iter_dtype is None:
-                iter_dtype = np.int32
-            _value_dict["iters"] = np.arange(_len, dtype=iter_dtype)
-
-        elif isinstance(iters, Number):
-            if _len != 1:
-                print("shape:", values.shape)
-                raise ValueError(
-                    f"""Need at least one iteration: {values}, {type(values)}, 
-                                {_is_scalar(values)}, {iters}, len:{_len},
-                                {values.shape}, {values.ndim}
-                                """
-                )
-            _value_dict["iters"] = np.array([iters], dtype=iter_dtype)
-
-        if _len != len(_value_dict["iters"]):
-            raise ErrorException("Not matching lengths")
-
-        self._value_dict = _value_dict
-        self._value_name = _value_name
-        self._len = _len
-        self._single_value = _single_value
-        self._keys = _keys
+            if is_scalar(val):
+                raise_if_len_not_match(1, n_elements, key)
+                val = np.asarray([val], dtype=dtype)
+
+            elif hasattr(val, "__array__"):
+                val = np.asarray(val, dtype=dtype)
+                if n_elements == 1 and len(val) != 1:
+                    val = np.reshape(val, (1,) + val.shape)
+
+                raise_if_len_not_match(len(val), n_elements, key)
+
+            elif isinstance(val, list):
+                pass
+            else:
+                val = [val]
+
+            value_dict[key] = val
+            keys.append(key)
+
+        self._value_dict = value_dict
+        self._value_name = main_value_name
+        self._len = n_elements
+        self._single_value = single_value
+        self._keys = keys
 
     @property
-    def iters(self):
+    def iters(self) -> Array:
         return self._value_dict["iters"]
 
     @property
-    def values(self):
+    def values(self) -> Array:
         return self._value_dict[self._value_name]
 
-    def __len__(self):
+    def __len__(self) -> int:
         return self._len
 
-    def __getitem__(self, key):
+    def __getitem__(self, key) -> Array:
+        # if its an int corresponding to an element not inside the dict,
+        # treat it as accessing a slice of a single element
+        if isinstance(key, int) and not key in self:
+            return self._get_slice(key)
+
+        # support slice syntax
+        if isinstance(key, slice):
+            return self._get_slice(key)
+
         return self._value_dict[key]
 
-    def __contains__(self, key):
+    def _get_slice(self, slce: slice) -> "History":
```

Comment:
```suggestion
    def _get_slice(self, slce: slice) -> History:
```

Reply:
Since it's the definition of "History", History (the object) does not exist yet, so within a class references to itself should be annotated as a string...

Reply:
🤮

Reply:
What's the oldest Python version we support? Because starting in 3.7 this should be solved by using `from __future__ import annotations` ([PEP 563](https://www.python.org/dev/peps/pep-0563/)):
```python3
In [1]: class X:
   ...:     def this(self) -> X:
   ...:         return self
   ...: 
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-1-47a2db087675> in <module>
----> 1 class X:
      2     def this(self) -> X:
      3         return self
      4 

<ipython-input-1-47a2db087675> in X()
      1 class X:
----> 2     def this(self) -> X:
      3         return self
      4 

NameError: name 'X' is not defined

In [2]: from __future__ import annotations

In [3]: class X:
   ...:     def this(self) -> X:
   ...:         return self
   ...: 
```

Reply:
We declare to support 3.7. I think we actually support ">=3.6" but we should check by running CI on 3.6...

about `from __future__ import annotations`...
it basically converts all annotations to string when python parses your code, before evaluating them.

We could enable it, but there is a [bit of a fight](https://github.com/samuelcolvin/pydantic/issues/2678) right now about whever this is actually the good `future`.
Some python developers are fighting the change and there's the competing [PEP649](https://www.python.org/dev/peps/pep-0649/) that would involve yet another change... They are expected to make a decision in the few weeks or so.

I'd like to wait to see what they settle for before adopting a certain `future`

Reply:
Alright, that sounds reasonable.